gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallContext.java

```diff
@@ -74,24 +71,26 @@
   /** Returns an empty instance with a null channel and default {@link CallOptions}. */
   public static GrpcCallContext createDefault() {
     return new GrpcCallContext(
-        null, CallOptions.DEFAULT, null, null, null, ImmutableMap.<String, List<String>>of());
+        null, CallOptions.DEFAULT, null, null, null, null, ImmutableMap.<String, List<String>>of());
   }
 
   /** Returns an instance with the given channel and {@link CallOptions}. */
   public static GrpcCallContext of(Channel channel, CallOptions callOptions) {
     return new GrpcCallContext(
-        channel, callOptions, null, null, null, ImmutableMap.<String, List<String>>of());
+        channel, callOptions, null, null, null, null, ImmutableMap.<String, List<String>>of());
   }
 
   private GrpcCallContext(
       Channel channel,
       CallOptions callOptions,
+      @Nullable Duration rpcTimeout,
```

Comment:
Storing `callOptions` and `rpcTimeout` separately duplicates essentially same configuration in two different fields. This is different from already existing `streamWaitTimeout` and  `streamIdleTimeout` because those are used as separate entities in `Watchdog`, while `rpcTimeout` is used solely to modify `callOptions`. If `callOptions` were not exposed themselves, that would be ok, but unfortunately it is not the case. Is there a really strong reason to store this value separately (I can see that having `getTimeout()` method makes unit testing much simpler, but it can be achieved without affecting the surface of the client)?

Reply:
This change has nothing to do with unit testing. It’s fixing a logic problem.

Although they have the same name, they are different things. The calloptions timeout is a deadline, which is an absolute timestamp that specifies when the request is no longer valid. An rpc timeout is a duration for how long a a single rpc can be active for. These concepts overlap when retries are disabled and the rpc is sent immediately after a timeout is set. However this is not the case when retries are enabled. As described in #561, calling apicontext.setTimeout will set an an absolute deadline for all attempts, which is inconsistent with behavior described in apicontext.setTimeout, which describes it as an rpc timeout.

In a future PR I will introduce apicontext.setTotalTimeout (or setDeadline), which will use calloptions.setTimeout.